### PR TITLE
26 implement processing step

### PR DESCRIPTION
**Pull request by led02**
Sorry for the big evil commit. However, this is a somewhat working version that

- Harvests data (from CITATION.cff, git history)
- Assembles them into a common codemeta.json
- Tracks the provenance of each entry
- Has some basic logging and configuration implemented
- Totally lacks documentation *blush*

Some stuff is lacking:

- Extensive tests
- Tutorials / docs
- Maybe some ADR

### CODE REVIEW

File: src/hermes/model/merge.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        'type': _check_types,
+        'path': _check_path,
+    }
+
+    def __init__(self, **filter):
+        self._filter = filter
+
+    def _check(self, key, filter, value):
+        if key in filter:
+            check = self.checks.get(key, lambda item, vaue: item in value)
```
Comment (sdruskat) [inline src/hermes/model/merge.py:49]:
```suggestion
            check = self.checks.get(key, lambda item, value: item in value)
```

Typo, probably fixing this leads to errors elsewhere.

File: src/hermes/model/path.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+        :return: The pyparsing.ParseResult.
+        """
+        return cls.path.parse_string(text)
+
+
+class ContextPath:
+    """
+    This class is used to access the different contexts.
+
+    On the one hand, the class allows you to define and manage pathes.
```
Comment (sdruskat) [inline src/hermes/model/path.py:36]:
```suggestion
    On the one hand, the class allows you to define and manage paths.
```

**typo:** :tada: for fixing 💋

File: src/hermes/model/path.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                raise TypeError(f'Invalid index type {type(index)} to set in {path.parent}.')
+
+            case _, _:
+                raise TypeError(f'Cannot handle target type {type(target)} to set {path}.')
+
+        return value
+
+    def resolve(self, target: list | dict, create: bool = False, query: t.Any = None) -> ('ContextPath', list | dict, 'ContextPath'):
+        """
+        Resolve a given path releative to a given target.
```
Comment (sdruskat) [inline src/hermes/model/path.py:266]:
```suggestion
        Resolve a given path relative to a given target.
```

**typo:** THanks for fixing.

**Review by sdruskat — APPROVED** [review]
LGTM, thanks 🌮 !

Couple of typos to be fixed, then please merge 🎉!

File: CITATION.cff [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  - given-names: Stephan
+    family-names: Druskat
+    email: stephan.druskat@dlr.de
+    affiliation: German Aerospace Center (DLR)
+    orcid: 'https://orcid.org/0000-0003-4925-7248'
+  - given-names: Jeffrey
+    family-names: Kelling
+    email: j.kelling@hzdr.de
+    affiliation: Helmholtz Zentrum Dresden-Rossendorf (HZDR)
+    orcid: 'https://orcid.org/0000-0003-1761-2591'
```
Comment (poikilotherm) [inline CITATION.cff:25]:
```suggestion
    orcid: 'https://orcid.org/0000-0003-1761-2591'
  - given-names: Oliver
    family-names: Bertuch
    email: o.bertuch@fz-juelich.de
    affiliation: Forschungszentrum Jülich GmbH (FZJ)
    orcid: 'https://orcid.org/0000-0002-2702-3419'  
```

**Comment by sdruskat** [comment]
Probably makes sense to merge #43 beforehand, to make sure that the REUSE workflow catches missing licenses, etc.

File: CITATION.cff [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# This CITATION.cff file was generated with cffinit.
+# Visit https://bit.ly/cffinit to generate yours today!
+
+cff-version: 1.2.0
+title: HERMES Aggregated Interface Script
```
Comment (jkelling) [inline CITATION.cff:5]: I am for having this file, we should lead by example. But, should this not be added in a separate PR? Seems unrelated to this one.

File: pyproject.toml [hunk 6]
```diff
@@ -24,7 +21,70 @@ myst-parser = "^0.17.2"
...
+000_cff = "hermes.commands.harvest.cff:harvest_cff"
+010_codemeta = "hermes.commands.harvest.codemeta:harvest_codemeta"
+020_git = "hermes.commands.harvest.git:harvest_git"
+
+[tool.poetry.plugins."hermes.preprocess"]
+000_cff = "hermes.commands.process.cff:add_name"
+020_git = "hermes.commands.process.git:flag_authors"
+
+[tool.hermes.harvest]
+000_git.enabled = false
```
[the remark this answers is not in the document]
Comment (led02) [reply pyproject.toml:35]: Will be changed as soon as we have configuration in place.

File: pyproject.toml [hunk 7]
```diff
@@ -24,7 +21,70 @@ myst-parser = "^0.17.2"
 haggis = "hermes.cli:haggis"
 
 [tool.poetry.plugins."hermes.harvest"]
-cff = "hermes.commands.harvest.cff:harvest_cff"
+000_cff = "hermes.commands.harvest.cff:harvest_cff"
+010_codemeta = "hermes.commands.harvest.codemeta:harvest_codemeta"
+020_git = "hermes.commands.harvest.git:harvest_git"
```
[the remark this answers is not in the document]
Comment (led02) [reply pyproject.toml:28]: KP22/678

[the remark this answers is not in the document]
Comment (led02) [reply pyproject.toml:28]: Will be changed as soon as we have configuration in place

File: pyproject.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (led02) [reply pyproject.toml:35]: No evaluated properly right now. Might be dropped

File: CITATION.cff (hunk 5 above)
Comment (led02) [reply CITATION.cff:5]: Good point. Will move it to another PR :)

File: pyproject.toml [hunk 8]
```diff
@@ -2,29 +2,116 @@
...
+sphinxext-opengraph = "^0.6.3"
 
 [tool.poetry.plugins.console_scripts]
 haggis = "hermes.cli:haggis"
 
 [tool.poetry.plugins."hermes.harvest"]
-cff = "hermes.commands.harvest.cff:harvest_cff"
+000_cff = "hermes.commands.harvest.cff:harvest_cff"
+010_codemeta = "hermes.commands.harvest.codemeta:harvest_codemeta"
+020_git = "hermes.commands.harvest.git:harvest_git"
```
Comment (jkelling) [inline pyproject.toml:52]: Maybe this i a point to bring this up: Why do we need this aliasing here? Why do we not just structure the module(s) in such a way that we do not the aliases, like plain old python modules?

I stumbled over this back when I added the git harverster. It confuses me, because this appears to needlessly add a requirement to manually replicate information in two places (here and in the filesystem/source tree) .. which seems especially weird as we are working on a convenience tool for meta-data handling here. I am not familiar with poetry ... so, is there a good reason for this?

Comment (led02) [reply pyproject.toml:52]: Well, these are not "aliases" but so-called entry points.

Python provides a method for extension of of your own code base - you can think of it like good old "plug-ins". This gives the opportunity of using an automated process of loading these Python functions without explicitly handling all the caveats that come with it. You can especially use these entry points to provide additional features from thrid-party Python packages. This would not be possible without entry points. On top, these entry points are a well-defined standard within Python.

The names here are in fact not very nice ... this will change as soon as we have a working configuration system. However, according to the specification of entry points, they need a name.

I hope that clearifies why this is needed.

File: src/hermes/commands/harvest/cff.py [hunk 9]
```diff
@@ -1,3 +1,4 @@
+import logging
```
Comment (poikilotherm) [inline src/hermes/commands/harvest/cff.py:1]: Move down to other imports

File: src/hermes/cli.py [hunk 10]
```diff
@@ -8,12 +8,40 @@
...
+from importlib import metadata
 
 import click
 
+from hermes import config
 from hermes.commands import workflow
+from hermes.config import configure, init_logging
+
+
+def log_header(header, summary=None):
```
Comment (poikilotherm) [inline src/hermes/cli.py:23]: Shouldn't this be living inside some util class / file?

File: src/hermes/commands/harvest/codemeta.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
+import glob
```
Comment (poikilotherm) [inline src/hermes/commands/harvest/codemeta.py:1]: Missing SPDX tags

File: src/hermes/commands/harvest/git.py [hunk 12]
```diff
@@ -0,0 +1,332 @@
+import datetime
```
Comment (poikilotherm) [inline src/hermes/commands/harvest/git.py:1]: Missing SPDX tags

File: src/hermes/commands/process/cff.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
+import logging
```
Comment (poikilotherm) [inline src/hermes/commands/process/cff.py:1]: Missing SPDX tags

File: src/hermes/commands/process/git.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
+import logging
```
Comment (poikilotherm) [inline src/hermes/commands/process/git.py:1]: Missing SPDX tags

File: src/hermes/commands/process/git.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    except ValueError:
+        audit_log.info("- Inconsistent data, skipping.")
+        return
+
+    for i, contributor in enumerate(author_path.get_from(data)):
+        query = {k: contributor[k] for k in _AUTHOR_KEYS if k in contributor}
+        author_key, target, path = author_path['*'].resolve(ctx._data, query=query)
+
+        if author_key._item == '*':
+            contributor['projectRole'] = 'Others'
```
[the remark this answers is not in the document]
Comment (poikilotherm) [reply src/hermes/commands/process/git.py:33]: How about https://credit.niso.org/?

File: src/hermes/config.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
+# TODO this file contains only dummy implementations which in most cases will lead to a crash...
```
Comment (poikilotherm) [inline src/hermes/config.py:1]: Missing SPDX tags

File: src/hermes/commands/workflow.py [hunk 17]
```diff
@@ -5,11 +5,15 @@
 # SPDX-FileContributor: Stephan Druskat
 # SPDX-FileContributor: Michael Meinel
 
+import json
+import logging
 from importlib import metadata
 
 import click
+import markdown as markdown
```
Comment (poikilotherm) [inline src/hermes/commands/workflow.py:13]: Can we remove this and remove the dependency?

File: src/hermes/model/merge.py [hunk 18]
```diff
@@ -0,0 +1,173 @@
+from hermes.model.path import ContextPath
```
Comment (poikilotherm) [inline src/hermes/model/merge.py:1]: Missing SPDX tags

File: src/hermes/model/path.py [hunk 19]
```diff
@@ -0,0 +1,372 @@
+import logging
```
Comment (poikilotherm) [inline src/hermes/model/path.py:1]: Missing SPDX tags

File: src/hermes/commands/process/git.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (sdruskat) [reply src/hermes/commands/process/git.py:33]: To be discussed in #87.

File: pyproject.toml [hunk 20]
```diff
@@ -33,24 +50,77 @@ optional = true
...
+
+[tool.poetry.plugins."hermes.preprocess"]
+000_cff = "hermes.commands.process.cff:add_name"
+020_git = "hermes.commands.process.git:flag_authors"
+
+
+[tool.hermes.harvest]
+000_git.enabled = false
```
Comment (sdruskat) [inline pyproject.toml:80]:
```suggestion
```

**Review by sdruskat — COMMENTED** [review]
Some final comments to resolve, then I think this is good to go.

**Review by jkelling — APPROVED** [review]
